### ISSUE TITLE
Modifying GALib for the objective function to do population based evaluation.

### ISSUE THREAD

**Issue by MichaelTettehK** [B0RJA/GAlib-mpi#6]
@B0RJA This is in no way related to your library. I just have a question about GAlib. I want to modifiy GAlib for the objective function to take an entire GAPopulation instead of GAGenome and return their fitnesses in a vector. Any help will be appreciated.

**Comment by B0RJA** [comment]
Genetic algorithms select the best adapted individuals to reproduce and transmit their characteristics to their offspring. For this reason, the objective function takes the genome of an individual as input and returns an score representing its adaptation. Group effects are often considered as competitions or championships in the core of the library but, to my knowledge, are also based on individual scores. I recommend you to check the description of the GAlib algorithm [1] or use a more modern genetic library like GAUL.

1. http://lancet.mit.edu/galib-2.4/Overview.html

**Comment by MichaelTettehK** [comment]
@B0RJA  Thanks. What I meant was to evaluate the entire population and return their fitnesses maybe as  a vector instead the current implementation which evaluates an individual and returns it raw fitness. Thanks again.

**Comment by B0RJA** [comment]
Check the function GAPopulation::DefaultEvaluator in the file ga-mpi/GAPopulation.C [1]. You will find a loop where each individual score is evaluated. You can replace this loop with a call to a function that takes the entire population as input. Honestly I do not see the interest of this function unless you want to consider group effects.

1. https://github.com/B0RJA/GAlib-mpi/blob/6c310756ea617b3ecbd0b0bba3f0348314ad77d5/ga-mpi/GAPopulation.C#L37

**Comment by MichaelTettehK** [comment]
@B0RJA Thanks. Basically, there is a library built on top of GALib that does grammatical evolution. Evaluating these individuals one at a time takes a long very time. But if I can get the entire population, the evaluator can evaluate all the individuals using multiple threads and return their fitnesses.

**Comment by B0RJA** [comment]
This was my problem and the reason of this repository. I implemented a threaded version of GALib using MPI to evaluate each individual in a different thread. Details of the parallel implementation can be see in the previously mentioned function and in some of the closed issues.

**Comment by MichaelTettehK** [comment]
@B0RJA  thanks.

**Comment by MichaelTettehK** [comment]
@B0RJA  I was trying to pass a function pointer from the GASState to the GAPopulation, so I can pass the population to this function as you suggested. But I am having issues, I think it has to do with the fact the the DefaultEvaluator is static. Any clues  how I can go about this?

**Comment by B0RJA** [comment]
I suggest you to start with the original version of GAlib [1] and create a repository with it and a functional example of your grammatical evolution library, evaluating the individuals one at a time. We continue talking when you have this working.

1. http://lancet.mit.edu/ga/dist/galib247.tgz

**Comment by MichaelTettehK** [comment]
@B0RJA Sorry for my long comments and thanks for your responses.
I'm using the original version of GAlib. The grammatical evolution (GE) library hasn't been released yet; I'm just a user, I don't think I can put it online. Basically, the GE is just acting as a mapper to map the genomes into a program, which is then evaluated and the fitness returned to be set as the GAGenome fitness. But because evaluating the programs one after the other is extremely slow, generating all the programs at from population will be much faster.

Basically, I want to have a function which takes the entire population for a generation and evaluates all of them and returns their fitness to be set. 
```
float * evaluatePopulation(GAPopulation &population) {
}
```
Getting this to work with GAlib is where I am struggling.

**Comment by B0RJA** [comment]
In that case I can not help you. I conclude suggesting that your approach is wrong. I think you do not need that function. Start understanding how GAlib works and connect it to the GE library at the correct point [1]. My implementation is a similar example that should help you achieve it.

1. https://github.com/B0RJA/GAlib-mpi/blob/6c310756ea617b3ecbd0b0bba3f0348314ad77d5/ga-mpi/GAPopulation.C#L37

**Comment by MichaelTettehK** [comment]
@B0RJA I went through your implementation. I couldn't find where you were calling the DefaultEvaluator method.
